Add an enhancement proposal for deleting child objects of policies (#67)
open-cluster-management-io/enhancements: enhancements/sig-policy/66-delete-policy-resources/README.md

  lastEvaluatedGeneration: 1
  relatedObjects:
  - compliant: Compliant
    object:
      apiVersion: v1
      kind: pods
      metadata:
        name: demo
        namespace: default
    properties:
      createdByPolicy: true //delete when parent configPolicy is deleted
    reason: Resource found as expected
```
### Scalability
The extra fields that will be needed to store data for this change should not impact scalability much, as we
would only add a few boolean fields to each `ConfigurationPolicy`. Deleting a policy will take a bit longer if
it is placed on a large amount of managed clusters, since the controller will now have to remove all created
objects from each managed cluster in addition to the `ConfigurationPolicy`. However, this would still be much
faster than it would take a user to clean up those resources otherwise - without this change, the easiest
way for users to clean up resources created by a policy would be to create another policy to delete the
objects after deleting the original policy, so this could end up helping with scalability because users
will not have to create as many policies on the hub cluster.

### Open Questions

> 

Reviewer: What behavior do we expect if the user chooses to edit the policy and change `remediationAction: enforce` to `inform`?  If the policy created resources and has `createdByPolicy: true`, it would seem the deletion of an `inform` policy would introduce a change to the cluster which is against the design for how `inform` works.  Should changing to inform disable the cleanup?

Also -- now that we have a cool namespace selector, what if a label on a namespace changes and the policy no longer applies to a particular namespace?  Will we realize our namespace matching is different and the resource needs to be cleaned up -- this is not a policy delete scenario, just a namespace matching change scenario.

Author: @gparvin Yeah, for the first question we would only delete the object if the remediationAction is set to enforce. The namespace mismatching is a good point but I don't think it applies to this enhancement since it's not something we need to worry about upon policy deletion. I think that we should open a separate issue to fix resources persisting after a namespace change (we could also include resources that have been renamed persisting with their old name in that second issue, since I think the solution would be similar).

is deleted. This will preserve existing behavior and make child object deletion
“opt-in.”

### User Stories

As a k8s admin, I want resources that are created by a policy to be cleaned
up automatically when I delete the policy that created them.

### Implementation Details/Notes/Constraints 

`removeChildrenOnDelete` will be at the spec level for the configurationPolicy
and thus the value it is set to will apply to all policy templates contained in
the configurationPolicy. This will make it simpler for users to set than having
the field be set at the template level. If users would like some objects to
remain on deletion and others to be deleted, they can separate those objects
into multiple configurationPolicies contained under the same parent policy.

ConfigurationPolicies already maintain a list of objects they manage in their
status field called `relatedObjects`. Since most of the information we would need
to track which objects need to be deleted already exists in this part of the status,
we propose adding a boolean field inside `status.relatedObjects.Object` that tells
the controller whether a resource needs to be deleted when the configurationPolicy

> is deleted. We would set this value when the configurationPolicy controller creates

Reviewer 2: ```suggestion
is deleted. We would set this value when the config policy controller creates
```

a resource during the handling of an “enforce” policy template. When processing
`spec.removeChildrenOnDelete`, we will set a finalizer on the configurationPolicy,
such as `policy.open-cluster-management.io/delete-related-objects`. When the policy
is deleted, the controller will perform cleanup logic on the objects in `relatedObjects`
that have been marked for deletion by the new status field. Once that cleanup is complete,
the finalizer will be removed and kubernetes will handle removing the configurationPolicy
from the cluster. Any resources without this value will persist on the cluster.

> Since `removeChildrenOnDelete` and `spec.removeChildrenOnDelete` are scoped to the

Reviewer 2: Did you mean this?
```suggestion
Since `spec.removeChildrenOnDelete` is scoped to the
```

> Since `removeChildrenOnDelete` and `spec.removeChildrenOnDelete` are scoped to the

Reviewer 3: i would not use boolean, a string will give you more extensibility in the future.

and that policy contains a configurationPolicy with `removeChildrenOnDelete: true`, the
object will be deleted on the old cluster and created on the new one. This means that
changing the placement of an `enforce` policy will no longer result in objects
persisting indefinitely on the old managed cluster.

### Risks and Mitigation

The main risk of adding an enhancement like this would be users deleting objects
that they want to remain on the cluster after they delete a policy. We plan to
mitigate this risk by making the `spec.removeChildrenOnDelete` field default to
`false`, so users will not delete anything without having to set that field first.

## Design Details

The `spec.removeChildrenOnDelete` field would look something like this:

```
spec:
  namespaceSelector:
    exclude:
    - kube-*
    include:
    - default
  object-templates:
  - complianceType: musthave
    objectDefinition:
      apiVersion: v1
      kind: Pod
      metadata:
        name: proposal-pod
      spec:
        containers:
        - image: nginx:1.18.0
          name: nginx
          ports:
          - containerPort: 80
  remediationAction: enforce
  severity: low

>   removeChildrenOnDelete: true //controller will remove proposal-pod

Reviewer 3: will all objects in your templates list be deleted, or you want to optionally delete some.

```

Here is what the configurationPolicy status would look like with the new field:

```
status:
  compliancyDetails:
  - Compliant: Compliant
    Validity: {}
    conditions:
    - lastTransitionTime: "2022-06-09T19:27:13Z"
      message: pods [demo] in namespace default found as specified, therefore this Object template is compliant
      reason: K8s `must have` object already exists
      status: "True"
      type: notification
  compliant: Compliant
  lastEvaluated: "2022-06-09T19:27:53Z"
  lastEvaluatedGeneration: 1
  relatedObjects:
  - compliant: Compliant
    object:
      apiVersion: v1
      kind: pods
      metadata:
        name: demo
        namespace: default
    properties:

>       createdByParent: true //delete when parent configPolicy is deleted

Reviewer 3: what if a user deletes the object then controller creates it, will it be set to true? You probably want to track the uid here since it could happen that the controller deletes one and another controller create a new one with new uid, and you may not want to delete the recreated object.

Reviewer 4: Tracking the uid could be good, although I think it's a weird situation.

It seems like it would only be necessary if the configpolicy creates an object, which is then deleted and recreated *before* the configpolicy tries to create it again. Because in this case the configpolicy didn't create the instance that it's still tracking.

But in general, whenever the configpolicy creates an object as opposed to editing it, I think we want to flag it like this. Regardless of whether that's when the configpolicy was first defined, or whether that happens later (eg after something deletes the object).

Reviewer 2: I'm wondering if we really need to distinguish whether a policy created the object in the first place or not. If a user sets the option to delete the objects after the policy is deleted, I think the objects should just be deleted regardless of who created it and when it was created.

For example, if you manually test out something on a managed cluster and then later make the policy do it on all managed clusters, it seems strange that the object would be deleted on all managed clusters except for the one that was manually tested it on.

Reviewer 4: These seem like additional reasons to consider an enum, we could have options like `AlwaysDelete`, `DeleteIfCreated`.

I think the "if created" logic originally came from wanting to be able to add this flag to existing policies, and do what we imagined users would want. But we could just give that option to the policy authors, maybe even per object-template for more control. We moved away from that because it is necessarily more complex for the user to need to define it in multiple places, but it would make the behavior more clear...

- E2E/integration tests to ensure that resources are being removed when policies are deleted and
resources persist when the spec field is not set

### Graduation Criteria

#### Alpha

1. The proposal is reviewed and accepted
2. Implementation is completed to support the primary goals
3. Test cases developed to demonstrate that the user stories are fulfilled

#### Beta

1. E2E/Integration tests complete

### Upgrade / Downgrade Strategy

The `spec.removeChildrenOnDelete` field being optional means that the controller will remain
backwards compatible with any older policies.

## Implementation History

N/A

## Drawbacks

- Users may accidentally delete resources if they are not familiar with how the new field works,
although they will not be able to do so unless they set the field without checking the documentation
on it.
- Since this will only remove objects that are specified in the policy yaml, objects created elsewhere
will not be deleted when the policy is deleted. This means the deletion will probably not work well
for removing operators, since they can create extra resources on the cluster that need to be cleaned
up.

> - Currently, the policy creation UI defaults to adding `clusterConditions` to the placementRule when

Reviewer 2: The UI isn't open-source. Could you please rephrase this to "some policy tooling" or similar?

creating a policy that make the placement only apply to healthy clusters. This means configurationPolicies
contained in the policy are deleted from a cluster that is unhealthy or hibernating, which would also
mean that with the changes detailed above, resources created by configurationPolicies on that cluster
would be deleted when the cluster is not running properly. The best solution to this would be to
change the policy creation UI to not add that cluster condition by default and then document that
changing the placement of a policy will cause objects to be removed from the old cluster if 
`spec.removeChildrenOnDelete` is set to true.

## Infrastructure Needed

No new infrastructure is needed, since all the necessary changes can take place in the existing

> configurationPolicy controller.

Reviewer 2: ```suggestion
config policy controller.
```

`ConfigurationPolicy` and adding a new field to the `ConfigurationPolicy` spec that allows
users to specify whether they want to delete any objects created by the `ConfigurationPolicy`.

## Motivation

The current implementation, which leaves old objects on the cluster, can make it a
pain to remove policies and confusing for users if they delete a policy without
keeping track of the objects that policy created. There is no way for users to
know which objects are left over after a policy is deleted, so those objects
will persist on the cluster and have potential to cause confusion. Even if users
know which objects need to be cleaned up after deleting a policy, it is time-consuming
to do so. By allowing users to specify that they would like to automatically delete
managed objects, we would make policy clean-up significantly simpler and easier.

### Goals

1. Object deletion: objects that are created by policies will be deleted when
the policy is deleted
2. Opt-in: add an optional new field to the `ConfigurationPolicy` spec and default
to leaving the object on the cluster upon policy deletion so that the behavior
of old policies is not changed

### Non-Goals

> 1. Deleting objects not created by the parent policy: we will not support

Reviewer 2: This seems to longer apply due to the `DeleteAll` option that was added to the proposal.

deleting objects that are not directly created by a `ConfigurationPolicy` when
that `ConfigurationPolicy` is deleted (i.e. existing resources that are edited
by the `ConfigurationPolicy`). We determined implementing this would create
more problems than it would solve, such as users accidentally deleting objects
that are created by the cluster or other policies.

## Proposal

> We would add a new field to the `ConfigurationPolicy` spec called

Reviewer 2: I would mention that this only applies when the `remediationAction` of `enforce` is set.

`spec.pruneObjectBehavior`. This field will be an enum with the options `DeleteAll`,
`DeleteIfCreated`, and `None`. When this field is set to `DeleteAll`, the controller
will automatically delete all objects referenced in the `ConfigurationPolicy` when
the `ConfigurationPolicy` is deleted. For `DeleteIfCreated`, the controller
will automatically delete any objects that it has created when the
`ConfigurationPolicy` is deleted and leave the objects that existed already when the
`ConfigurationPolicy` was created. When it is set to `None` or not set, the
controller will leave any created objects intact once the `ConfigurationPolicy`
is deleted. This will preserve existing behavior and make child object deletion
“opt-in.”

### User Stories

> As a k8s admin, I want resources that are created by a policy to be cleaned

Reviewer 2: ```suggestion
As a k8s admin, I want objects that are created by a policy to be cleaned
```

up automatically when I delete the policy that created them.

### Implementation Details/Notes/Constraints 

`pruneObjectBehavior` will be at the spec level for the `ConfigurationPolicy`
and thus the value it is set to will apply to all object templates contained in
the `ConfigurationPolicy`. This will make it simpler for users to set than having
the field be set at the template level. If users would like some objects to
remain on deletion and others to be deleted, they can separate those objects
into multiple `ConfigurationPolicy` objects contained under the same parent policy.

`ConfigurationPolicy` objects already maintain a list of objects they manage in their
status field called `relatedObjects`. Since most of the information we would need
to track which objects need to be deleted already exists in this part of the status,
we propose adding a boolean field inside `status.relatedObjects.Object` that tells

> the controller whether a resource needs to be deleted when the `ConfigurationPolicy`

Reviewer 2: ```suggestion
the controller whether the object needs to be deleted when the `ConfigurationPolicy`
```

is deleted. We would set this value when the config policy controller creates

> a resource during the handling of an “enforce” policy template. Additionally, we will

Reviewer 2: ```suggestion
an object during the handling of an “enforce” policy template. Additionally, we will
```

> add the object's UID to the status field so that we can determine whether the object

Reviewer 4: It will be used to verify the identity of the object to be deleted. Say a config policy creates `pod/foo`. Then, between evaluation loops, something else deletes and recreates that pod. In this case, if the behavior is `DeleteIfCreated`, we don't want the policy to delete the pod. In some sense, the "responsibility" for the pod has been passed to whatever else recreated it.

With just the name, we can't detect this situation. With a UID, we can.

> add the object's UID to the status field so that we can determine whether the object

Reviewer 2: Sounds good. @willkutler could you please add this to the doc?

is owned by the `ConfigurationPolicy` or not. 

When processing `spec.pruneObjectBehavior`, we will set a finalizer on the `ConfigurationPolicy`,
such as `policy.open-cluster-management.io/delete-related-objects`. When the policy
is deleted, the controller will perform cleanup logic on the objects in `relatedObjects`
that have been marked for deletion by the new status field. Once that cleanup is complete,
the finalizer will be removed and kubernetes will handle removing the `ConfigurationPolicy`

> from the cluster. Any resources without this value will persist on the cluster.

Reviewer 2: ```suggestion
from the cluster. Any objects without this value will persist on the cluster.
```

  compliancyDetails:
  - Compliant: Compliant
    Validity: {}
    conditions:
    - lastTransitionTime: "2022-06-09T19:27:13Z"
      message: pods [demo] in namespace default found as specified, therefore this Object template is compliant
      reason: K8s `must have` object already exists
      status: "True"
      type: notification
  compliant: Compliant
  lastEvaluated: "2022-06-09T19:27:53Z"
  lastEvaluatedGeneration: 1
  relatedObjects:
  - compliant: Compliant
    object:
      apiVersion: v1
      kind: pods
      metadata:
        name: demo
        namespace: default
    properties:
      createdByPolicy: true //delete when parent configPolicy is deleted
      uid: 1234567890
    reason: Resource found as expected
```
### Scalability
The extra fields that will be needed to store data for this change should not impact scalability much, as we
would only add a few boolean fields to each `ConfigurationPolicy`. Deleting a policy will take a bit longer if
it is placed on a large amount of managed clusters, since the controller will now have to remove all created
objects from each managed cluster in addition to the `ConfigurationPolicy`. However, this would still be much

> faster than it would take a user to clean up those resources otherwise - without this change, the easiest

Reviewer 2: ```suggestion
faster than it would take a user to clean up those objects otherwise - without this change, the easiest
```

> way for users to clean up resources created by a policy would be to create another policy to delete the

Reviewer 2: ```suggestion
way for users to clean up objects created by a policy would be to create another policy to delete the
```

objects after deleting the original policy, so this could end up helping with scalability because users
will not have to create as many policies on the hub cluster.

### Open Questions

### Test Plan

- Unit tests can be implemented on some of the helper functions that we add to the controller, but

> will not be able to test whether resources are being removed properly.

Reviewer 2: ```suggestion
will not be able to test whether objects are being removed properly.
```

> - E2E/integration tests to ensure that resources are being removed when policies are deleted and

Reviewer 2: ```suggestion
- E2E/integration tests to ensure that objects are being removed when policies are deleted and
```

> resources persist when the spec field is not set

Reviewer 2: ```suggestion
objects persist when the spec field is not set
```

### Graduation Criteria

#### GA

1. The proposal is reviewed and accepted
2. Implementation is completed to support the primary goals
3. Test cases developed to demonstrate that the user stories are fulfilled
4. E2E/Integration tests complete

### Upgrade / Downgrade Strategy

The `spec.pruneObjectBehavior` field being optional means that the controller will remain
backwards compatible with any older policies.

## Implementation History

N/A

## Drawbacks

> - Users may accidentally delete resources if they are not familiar with how the new field works,

Reviewer 2: ```suggestion
- Users may accidentally delete objects if they are not familiar with how the new field works,
```

although they will not be able to do so unless they set the field without checking the documentation
on it.
- Since this will only remove objects that are specified in the policy yaml, objects created elsewhere
will not be deleted when the policy is deleted. This means the deletion will probably not work well

> for removing operators, since they can create extra resources on the cluster that need to be cleaned

Reviewer 2: ```suggestion
for removing operators, since they can create extra objects on the cluster that need to be cleaned
```

up.
- Currently, some policy tooling defaults to adding `clusterConditions` to the placementRule when
creating a policy that make the placement only apply to healthy clusters. This means `ConfigurationPolicy` objects
contained in the policy are deleted from a cluster that is unhealthy or hibernating, which would also

> mean that with the changes detailed above, resources created by `ConfigurationPolicy` objects on that cluster

Reviewer 2: ```suggestion
mean that with the changes detailed above, objects created by `ConfigurationPolicy` objects on that cluster
```
